### PR TITLE
Add support for OAuth2 and OpenID Connect providers. 

### PR DESCRIPTION
**Pull request by EreMaijala**
This PR adds support for VuFind to act as an identity provider for external services using OpenID Connect (OIDC).

This can be very useful e.g. when VuFind provides access to ebooks etc. with a provider that needs to have some information about the user, such as birthdate, to control access. OIDC gives the user control over release of their information, which helps also with the GDPR requirements.

The PR does not currently include any functionality to remember and/or revoke user consents, so users will need to allow access every time. Consent management could be added in a follow-up task.

A large part of the code is plumbing for league/oauth2-server. The implementation loosely follows oauth2-server-example and is also inspired by Drupal's simple_oauth module.

[idp-oidc-tester](https://medium.com/application-security/openid-connect-tester-for-developers-63427324faea) is a very useful tool for testing the OIDC provider. Example to run it with docker: `docker run --rm --name idp-oidc-tester -p 8080:80 registry.gitlab.com/guenoledc-perso/idp-oidc-tester:latest` (make sure that the service in Docker can access token, jwks and user info endpoints; with Mac use e.g. http://docker.for.mac.localhost/vufind2/OAuth2/token).

TODO:
- [x] Add a scope for patron's blocks
- [x] Tests
- [x] Complete birthdate support in a separate PR (#2502)
- [x] [Wiki page](https://vufind.org/wiki/configuration:oauth2_oidc)
- [x] Identify and fix any places where the code could be cleaned up
- [x] Add and test Postgresql support

### CODE REVIEW

File: languages/en.ini [hunk 1]
```diff
@@ -826,6 +826,23 @@ Number = "Number"
 number_decimal_point = "."
 number_thousands_separator = ","
 OAI Server = "OAI Server"
+oauth2_allow_access = "Allow Access"
```
Comment (demiankatz) [inline languages/en.ini]: I wonder if we should use a more generic prefix than oauth2 -- I assume that this text would be the same for any sort of external permission system, whether the technology is oauth2 or something else. Might be more future-proof to not tie the translation strings to a specific protocol version. Maybe "external_auth_..."?

**Review by demiankatz — COMMENTED** [review]
I'll give this a deeper review later in the week, but I had five minutes just now and started with one minor comment... :-)

File: languages/en.ini (hunk 1 above)
Comment (EreMaijala) [reply languages/en.ini]: Makes sense, I'll revise. :)

File: config/vufind/OAuth2Server.yaml [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+  family_name: lastname
+  email: email
+  birthdate: birthdate
+  locale: last_language
+  phone: phone
+  address: address_json # special field including all address fields
+# Grants configuration
+Grants:
+  # Life time for an authorization code (should be kept short). Default is 1 minute.
+  authCodeLifeTime: PT1M
```
Comment (demiankatz) [inline config/vufind/OAuth2Server.yaml:155]: I'm not familiar with the `PT...` syntax. Does this need to be explained in the comments?

File: languages/en.ini [hunk 3]
```diff
@@ -410,8 +410,25 @@ export_selected_favorites = "Export Selected Items"
...
 export_success = "Export Ready"
 export_unsupported_format = "Unsupported Export Format"
+external_auth_access_heading = "Request for Access"
+external_auth_access_login_message = "An external application is requesting access to your data. Please log in to review and authorize the request."
+external_auth_allow_access = "Allow Access"
+external_auth_deny_access = "Deny Access"
 external_auth_heading = "Access to licensed material"
+external_auth_ils_prompt = "Some of the requested information is retrieved from your library catalog profile."
 external_auth_login_message = "Login to access licensed material"
+external_auth_prompt_html = "is requesting the following access rights:"
```
Comment (demiankatz) [inline languages/en.ini:427]: This seems to be missing the client name token.

File: module/VuFind/src/VuFind/ILS/Driver/Demo.php [hunk 4]
```diff
@@ -904,7 +907,8 @@ public function getMyProfile($patron)
             'phone'           => '1900 CALL ME',
             'mobile_phone'    => '1234567890',
             'group'           => 'Library Staff',
-            'expiration_date' => 'Someday'
+            'expiration_date' => 'Someday',
+            'birthdate'       => $birthDate->format('Y-m-d')
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/ILS/Driver/Demo.php]: I don't think birthdate is currently part of the driver spec. Would it make sense to add this as a separate PR?

File: module/VuFind/src/VuFind/ILS/Driver/SierraRest.php [hunk 5]
```diff
@@ -554,6 +554,7 @@ public function getMyProfile($patron)
             'address1' => $address,
             'zip' => $zip,
             'city' => $city,
+            'birthdate' => $result['birthDate'] ?? '',
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/ILS/Driver/SierraRest.php]: Is the format of this guaranteed across all Sierra installations?

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks for the progress here. I still haven't completed a full review yet, but I found a few more things to comment on as I continued chipping away at the list of files.

File: config/vufind/OAuth2Server.yaml (hunk 2 above)
Comment (EreMaijala) [reply config/vufind/OAuth2Server.yaml:155]: Comment expanded.

File: languages/en.ini (hunk 3 above)
Comment (EreMaijala) [reply languages/en.ini:427]: The token was a leftover. The client name is in the heading.

File: module/VuFind/src/VuFind/ILS/Driver/Demo.php (hunk 4 above)
Comment (EreMaijala) [reply module/VuFind/src/VuFind/ILS/Driver/Demo.php]: Yes, and that was my intention. I expanded the TODO item.

File: module/VuFind/src/VuFind/ILS/Driver/SierraRest.php (hunk 5 above)
Comment (EreMaijala) [reply module/VuFind/src/VuFind/ILS/Driver/SierraRest.php]: Yes. See https://sandbox.iii.com/iii/sierra-api/swagger/index.html#!/patrons/Get_a_list_of_patrons_get_1. It's been stable across API versions.

File: module/VuFind/src/VuFind/Controller/OAuth2ControllerFactory.php [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+
+    /**
+     * Return a factory function for creating the authorization server.
+     *
+     * @return callable
+     */
+    protected function getAuthorizationServerFactory(): callable
+    {
+        return function (?string $clientId): AuthorizationServer {
+            if (!($keyPath = $this->oauth2Config['Server']['privateKeyPath'] ?? '')
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Controller/OAuth2ControllerFactory.php]: Adding a `getOauth2ServerSetting` method might make the code a little more concise, since you could centralize the value checking and exception throwing to a single place.

File: module/VuFind/src/VuFind/Controller/OAuth2ControllerFactory.php [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+    ) {
+        if (!empty($options)) {
+            throw new \Exception('Unexpected options sent to factory.');
+        }
+        $yamlReader = $container->get(\VuFind\Config\YamlReader::class);
+        $oauth2Config = $yamlReader->get('OAuth2Server.yaml');
+
+        $this->container = $container;
+        $this->oauth2Config = $oauth2Config;
+        $tablePluginManager = $this->container
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Controller/OAuth2ControllerFactory.php]: Any particular reason why you use `$this->container` here but `$container` everywhere else in this method?

File: module/VuFind/src/VuFind/Controller/OAuth2Controller.php [hunk 8]
```diff
@@ -0,0 +1,438 @@
...
+                Psr7ServerRequest::fromLaminas($this->getRequest())
+            );
+        } catch (OAuthServerException $exception) {
+            return $this->convertOAuthServerExceptionToResponse($exception);
+        } catch (\Exception $e) {
+            return $this->handleException('Authorization request', $e);
+        }
+
+        if ($this->formWasSubmitted('allow') || $this->formWasSubmitted('deny')
+        ) {
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Controller/OAuth2Controller.php]: Does this line really need to be wrapped?

File: module/VuFind/src/VuFind/Controller/OAuth2Controller.php [hunk 9]
```diff
@@ -0,0 +1,438 @@
...
+            if (!$this->csrf->isValid($this->getRequest()->getPost()->get('csrf'))) {
+                throw new \VuFind\Exception\BadRequest(
+                    'error_inconsistent_parameters'
+                );
+            }
+
+            // Store OpenID nonce in the access token table so that it can be
+            // retrieved for token or user info action:
+            $this->accessTokenTable
+                ->storeNonce($user->id, $laminasRequest->getQuery('nonce'));
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Controller/OAuth2Controller.php:233]: Do we need to worry about the nonce being absent?

File: module/VuFind/src/VuFind/Controller/OAuth2ControllerFactory.php [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+    public function __invoke(
+        ContainerInterface $container,
+        $requestedName,
+        array $options = null
+    ) {
+        if (!empty($options)) {
+            throw new \Exception('Unexpected options sent to factory.');
+        }
+        $yamlReader = $container->get(\VuFind\Config\YamlReader::class);
+        $oauth2Config = $yamlReader->get('OAuth2Server.yaml');
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Controller/OAuth2ControllerFactory.php]: You could save a line of code by assigning directly to `$this->oauth2Config` here and then changing the access to `$oauth2Config` below.

File: module/VuFind/sql/mysql.sql [hunk 11]
```diff
@@ -346,3 +346,21 @@ CREATE TABLE `auth_hash` (
 ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE utf8mb4_unicode_ci;
 /*!40101 SET character_set_client = @saved_cs_client */;
 
+--
+-- Table structure for table `access_tokens`
+--
+
+/*!40101 SET @saved_cs_client     = @@character_set_client */;
+/*!40101 SET character_set_client = utf8mb4 */;
+CREATE TABLE `access_token` (
```
Comment (demiankatz) [inline module/VuFind/sql/mysql.sql:382]: We should make sure that PostgreSQL support is on the TODO list -- though I'm fine with doing that toward the end of the project so we don't have to spend time porting anything that changes along the way.

File: module/VuFind/src/VuFind/OAuth2/Repository/ScopeRepository.php [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+     * Return information about a scope.
+     *
+     * @param string $identifier The scope identifier
+     *
+     * @return ScopeEntityInterface|null
+     */
+    public function getScopeEntityByIdentifier($identifier)
+    {
+        if (!($config = $this->oauth2Config['Scopes'][$identifier] ?? [])) {
+            $config['hidden'] = true;
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/OAuth2/Repository/ScopeRepository.php]: Should this actually be `$config = ['hidden' => true]` in case we get a non-array value out of the configuration, or is that impossible?

File: module/VuFind/src/VuFind/Controller/OAuth2ControllerFactory.php [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+            $encryptionKey
+                = trim($this->oauth2Config['Server']['encryptionKey'] ?? '');
+            if (!$encryptionKey) {
+                throw new \Exception(
+                    'Server/encryptionKey missing from OAuth2Server.yaml'
+                );
+            }
+
+            $server = new AuthorizationServer(
+                new ClientRepository($this->oauth2Config),
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Controller/OAuth2ControllerFactory.php]: Should we have separate getters (or perhaps properties containing class names) for the various repositories and other VuFind-specific classes so it's easier to locally override them?

File: module/VuFind/src/VuFind/OAuth2/Entity/ScopeEntity.php [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+    public function __construct(array $config)
+    {
+        foreach (['identifier'] as $required) {
+            if (!isset($config[$required])) {
+                throw new \Exception("OAuth2 scope config missing '$required'");
+            }
+        }
+        $this->setIdentifier($config['identifier']);
+        $this->setDescription($config['description'] ?? '');
+        $this->setILSNeeded(boolval($config['ils'] ?? false));
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/OAuth2/Entity/ScopeEntity.php]: Any strong reason to use `boolval` instead of `(bool)` typecasting? Nothing wrong with it, but I feel like it's not how we usually force values to booleans in the code base.

**Review by demiankatz — COMMENTED** [review]
I don't claim to know a lot of the Oauth2/OpenID-specific details, but I gave this a general read-through and had a few (mostly minor stylistic) suggestions.

File: module/VuFind/src/VuFind/Controller/OAuth2Controller.php (hunk 9 above)
Comment (EreMaijala) [reply module/VuFind/src/VuFind/Controller/OAuth2Controller.php:233]: storeNonce allows it to be null as well, which would be fine. I'll update the comment to clarify that.

File: module/VuFind/src/VuFind/Controller/OAuth2Controller.php (hunk 8 above)
Comment (EreMaijala) [reply module/VuFind/src/VuFind/Controller/OAuth2Controller.php]: No, fixed.

File: module/VuFind/src/VuFind/Controller/OAuth2ControllerFactory.php (hunk 7 above)
Comment (EreMaijala) [reply module/VuFind/src/VuFind/Controller/OAuth2ControllerFactory.php]: No, fixed.

File: module/VuFind/src/VuFind/Controller/OAuth2ControllerFactory.php (hunk 10 above)
Comment (EreMaijala) [reply module/VuFind/src/VuFind/Controller/OAuth2ControllerFactory.php]: Fixed.

File: module/VuFind/sql/mysql.sql (hunk 11 above)
Comment (EreMaijala) [reply module/VuFind/sql/mysql.sql:382]: I added a TODO entry and will add it.

File: module/VuFind/src/VuFind/OAuth2/Repository/ScopeRepository.php (hunk 12 above)
Comment (EreMaijala) [reply module/VuFind/src/VuFind/OAuth2/Repository/ScopeRepository.php]: Yes, that's better. Fixed.

File: module/VuFind/src/VuFind/Controller/OAuth2ControllerFactory.php (hunk 13 above)
Comment (EreMaijala) [reply module/VuFind/src/VuFind/Controller/OAuth2ControllerFactory.php]: I added some getters and then decided that it's better to actually utilize the service manager, so now it all goes through it. I also split some methods to smaller parts for easier overriding.

File: module/VuFind/src/VuFind/OAuth2/Entity/ScopeEntity.php (hunk 14 above)
Comment (EreMaijala) [reply module/VuFind/src/VuFind/OAuth2/Entity/ScopeEntity.php]: No, changed.

File: module/VuFind/src/VuFind/Controller/OAuth2ControllerFactory.php (hunk 13 above)
Comment (demiankatz) [reply module/VuFind/src/VuFind/Controller/OAuth2ControllerFactory.php]: Thanks, @EreMaijala, this is definitely cleaner now. I did momentarily wonder if we should have an Oauth2-specific plugin manager to avoid putting so many narrow-purpose definitions into the top-level service manager, but given that these classes don't have a common base class or interface, I think what you've done makes the most sense. If we wanted to separate these concerns from the rest of the code, maybe it would actually make the most sense to factor them to a separate VuFind\Oauth2 Laminas module... but I don't think that's something we need to contemplate right now, unless it appeals to you.

Comment (EreMaijala) [reply module/VuFind/src/VuFind/Controller/OAuth2ControllerFactory.php]: I thought about a plugin manager, but it didn't feel appropriate. A separate module is worth a thought, but I feel it would be so intertwined with the base module that it gets complicated easily, particularly if the functionality is expanded in the future so that it'd be possible to remember and manage access grants.

Comment (demiankatz) [reply module/VuFind/src/VuFind/Controller/OAuth2ControllerFactory.php]: Agreed -- worth consideration, but not an urgent priority right now. We can see how this evolves and refactor later if justified.

File: config/vufind/OAuth2Server.yaml [hunk 15]
```diff
@@ -0,0 +1,159 @@
+# OAuth2 Server Configuration
```
Comment (demiankatz) [inline config/vufind/OAuth2Server.yaml:1]: Maybe we should provide a little more context here, like a link to a wiki page, so somebody curious about the purpose of this file can quickly figure it out.

File: config/vufind/OAuth2Server.yaml [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+
+# Authorization server configuration
+Server:
+  # Create keys e.g. as above and point the following setting to it. The paths can be
+  # absolute or relative to the config directory.
+  privateKeyPath: "oauth2_private.key"
+  publicKeyPath: "oauth2_public.key"
+  # Encryption key used to encrypt payloads. Make sure it is random and long enough
+  # (at least 32 characters).
+  encryptionKey: ""
```
Comment (demiankatz) [inline config/vufind/OAuth2Server.yaml:22]: Since we're defaulting to blank here, will the software fail if the user does not fill it in? I agree that it's a good idea not to provide a default key, since that could compromise security... just want to be sure there's a check to ensure we don't go keyless by accident.

Comment (demiankatz) [reply config/vufind/OAuth2Server.yaml:22]: (Same for the salt below).

**Review by demiankatz — COMMENTED** [review]
I still have more to review, but a few comments from progress I made today...

File: module/VuFind/src/VuFindTest/Integration/MinkTestCase.php [hunk 17]
```diff
@@ -58,6 +59,13 @@ abstract class MinkTestCase extends \PHPUnit\Framework\TestCase
      */
     protected $modifiedConfigs = [];
 
+    /**
+     * Modified yaml configurations
+     *
+     * @var array
+     */
+    protected $modifiedYamlConfigs = [];
```
Comment (demiankatz) [inline module/VuFind/src/VuFindTest/Integration/MinkTestCase.php]: This YAML functionality is a nice addition; maybe we should commit it separately from the rest. I can merge it early to dev if you like, just to reduce files in this PR.

File: module/VuFind/src/VuFind/Controller/OAuth2Controller.php [hunk 18]
```diff
@@ -0,0 +1,439 @@
...
+    }
+
+    /**
+     * Get a JSON response from an array of data
+     *
+     * @param array $data Data to encode
+     *
+     * @return Response
+     */
+    protected function getJsonResponse(array $data): Response
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Controller/OAuth2Controller.php]: Might it make sense to put some of these utility methods like getJsonResponse and addCorsHeaders into a trait for reuse? I'd eventually like to clean up the existing controller class hierarchy by moving more logic out of the base class and into traits... maybe this is an opportunity to begin to move in that direction.

File: module/VuFind/src/VuFind/OAuth2/Repository/ScopeRepository.php [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+     * Given a client, grant type and optional user identifier validate the set of
+     * scopes requested are valid and optionally append additional scopes or remove
+     * requested scopes.
+     *
+     * @param ScopeEntityInterface[] $scopes         Scopes
+     * @param string                 $grantType      Grant type
+     * @param ClientEntityInterface  $clientEntity   Client
+     * @param null|string            $userIdentifier User ID
+     *
+     * @return ScopeEntityInterface[]
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/OAuth2/Repository/ScopeRepository.php:89]: Do we need an unused parameters annotation here?

File: config/vufind/OAuth2Server.yaml (hunk 16 above)
Comment (EreMaijala) [reply config/vufind/OAuth2Server.yaml:22]: `getOAuth2ServerSetting` will ensure that there is at least something in the key. It doesn't ensure the key is great, but I think it's okay to leave that to the administrator. Also the grants that encrypt something will throw if the key is empty. And yes, leaving the default empty helps avoid [trouble](﻿https://it.slashdot.org/story/22/08/22/227245/hyundai-uses-example-keys-for-encryption-system).

Comment (EreMaijala) [reply config/vufind/OAuth2Server.yaml:22]: And salt is also retrieved via `getOAuth2ServerSetting` that ensures it's set.

File: module/VuFind/src/VuFindTest/Integration/MinkTestCase.php (hunk 17 above)
Comment (EreMaijala) [reply module/VuFind/src/VuFindTest/Integration/MinkTestCase.php]: Sure, please do!

**Comment by demiankatz** [comment]
Okay, @EreMaijala, I think that completes my review... but on the understanding that I'm not familiar with the OAuth2 libraries and did not study that code in great depth -- I focused mainly on the changes to existing code and skimmed the rest. If we merge, we should check the coverage report and make sure everything new is fully tested (if you haven't already looked at coverage reports locally, of course). We also need to remember to adapt all this when we eventually merge it from dev into #2233.

File: module/VuFind/src/VuFindTest/Integration/MinkTestCase.php (hunk 17 above)
Comment (demiankatz) [reply module/VuFind/src/VuFindTest/Integration/MinkTestCase.php]: Done: commit c906c31584dcdb8b33eb3bfc804159ee8e89b30d.

File: module/VuFind/src/VuFind/OAuth2/Repository/ScopeRepository.php (hunk 19 above)
Comment (EreMaijala) [reply module/VuFind/src/VuFind/OAuth2/Repository/ScopeRepository.php:89]: Yes, added.

File: module/VuFind/src/VuFind/Controller/OAuth2Controller.php (hunk 18 above)
Comment (EreMaijala) [reply module/VuFind/src/VuFind/Controller/OAuth2Controller.php]: Makes sense. Trait and accompanying tests added.

**Comment by EreMaijala** [comment]
@demiankatz Thanks! All issues should be resolved now. I'll try to verify the test coverage once more.

**Comment by EreMaijala** [comment]
@demiankatz Code coverage for unit tests looks good to me.

**Review by demiankatz — APPROVED** [review]
Thanks, @EreMaijala!
